cpp/tests/copying/segmented_gather_list_tests.cpp

```diff
@@ -0,0 +1,446 @@
+      std::vector<bool> expected_valids = {false, true, true, false, false, true};
+
+      LCW<T> expected{{{{LCW<T>{30}},
+                        {{{10, 20}, valids}},
+                        {{{10, 20}, valids}},
+                        {LCW<T>{30}},
+                        {LCW<T>{30}},
+                        {{{10, 20}, valids}}},
+                       expected_valids.begin()},
+                      {{{LCW<T>{30}},
+                        {{10, 20, 30}},
+                        {{10, 20, 30}},
+                        {LCW<T>{30}},
+                        {LCW<T>{30}},
+                        {{{20, 30}, valids}, {62, 72, 82}}},
+                       expected_valids.begin()}};
+      CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(expected, result->view());
+    }
+  }
+}
+
+using SegmentedGatherTestString = SegmentedGatherTest<int>;
+TEST_F(SegmentedGatherTestString, StringGather)
+{
+  using T = cudf::string_view;
+  // List<T>
+  LCW<T> list{{"a", "b", "c", "d"}, {"1", "2", "3", "4"}, {"x", "y", "z"}};
+  LCW<int8_t> gather_map{{0, 1, 3, 2}, {1, 0, 3, 2}, LCW<int8_t>{}};
+  LCW<T> expected{{"a", "b", "d", "c"}, {"2", "1", "4", "3"}, LCW<T>{}};
+
+  auto result =
+    cudf::lists::detail::segmented_gather(lists_column_view{list}, lists_column_view{gather_map});
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected, result->view());
+}
+
+using SegmentedGatherTestInt = SegmentedGatherTest<int>;
+TEST_F(SegmentedGatherTestInt, Fails)
+{
+  using T = int32_t;
+  // List<T>
+  LCW<T> list{{1, 2, 3, 4}, {5}, {6, 7}, {8, 9, 10}};
+  LCW<int8_t> size_mismatch_map{{3, 2, 1, 0}, {0}, {0, 1}};
+  cudf::test::fixed_width_column_wrapper<int> nonlist_map0{1, 2, 0, 1};
+  cudf::test::strings_column_wrapper nonlist_map1{"1", "2", "0", "1"};
+  LCW<cudf::string_view> nonlist_map2{{"1", "2", "0", "1"}};
+
+  CUDF_EXPECT_THROW_MESSAGE(
+    cudf::lists::detail::segmented_gather(lists_column_view{list}, lists_column_view{nonlist_map0}),
+    "lists_column_view only supports lists");
+
+  CUDF_EXPECT_THROW_MESSAGE(
+    cudf::lists::detail::segmented_gather(lists_column_view{list}, lists_column_view{nonlist_map1}),
+    "lists_column_view only supports lists");
+
+  CUDF_EXPECT_THROW_MESSAGE(
+    cudf::lists::detail::segmented_gather(lists_column_view{list}, lists_column_view{nonlist_map2}),
+    "Gather map should be list column of index type");
+
+  auto valids = cudf::test::make_counting_transform_iterator(0, [](auto i) { return i % 2 == 0; });
+  LCW<int8_t> nulls_map{{{3, 2, 1, 0}, {0}, {0}, {0, 1}}, valids};
+  CUDF_EXPECT_THROW_MESSAGE(
+    cudf::lists::detail::segmented_gather(lists_column_view{list}, lists_column_view{nulls_map}),
+    "Gather map contains nulls");
+
+  CUDF_EXPECT_THROW_MESSAGE(cudf::lists::detail::segmented_gather(
+                              lists_column_view{list}, lists_column_view{size_mismatch_map}),
+                            "Gather map and list column should be same size");
+}
```

Comment:
If we're going to support a sliced gather_map, add a test here for that.

Reply:
added. (originally it was only sliced value_column. sliced gather map support was also added in last commit)